### PR TITLE
WebsocketResponse future now fails if the connection failed #21051

### LINKED ISSUE #21051: Http().webSocketClientFlow materialized value doesn't complete when TCP cannot connect
Please, try the following client application with akka-stream-2.4.8

```scala
import akka.actor.ActorSystem
import akka.http.scaladsl._
import akka.http.scaladsl.model.ws._
import akka.stream.ActorMaterializer
import akka.stream.scaladsl._
import scala.concurrent.Future

object Issue extends App {
  
  implicit val system = ActorSystem()
  implicit val mat = ActorMaterializer()
  implicit val ec = system.dispatcher
  
  val outgoing = 
    Source.single(TextMessage("hello world!"))

  val webSocketFlow: Flow[Message, Message, Future[WebSocketUpgradeResponse]] =
    Http().webSocketClientFlow(WebSocketRequest("ws://127.0.0.1:9999/no/server/here"))

  val incoming =
    Sink.foreach[Message] {
      case message: TextMessage.Strict =>
        println(message.text)
    }

  // upgradeResponse is a Future[WebSocketUpgradeResponse]
  // and it's expected to complete with success or failure
  val (upgradeResponse, closed) =
    outgoing
      .viaMat(webSocketFlow)(Keep.right)
      .toMat(incoming)(Keep.both)
      .run()
  
  // BUT it doesn't complete with failure as it should
  upgradeResponse.onFailure {
    case t: Throwable => 
      println(s"Connection failed : ${t.getMessage}")
  }
}
```

As I can read in the official [Client-Side WebSocket Support](http://doc.akka.io/docs/akka/2.4.8/scala/http/client-side/websocket-support.html#websocketclientflow) documentation, I'm just holding the ``upgradeResponse`` materialized value of the ``websocketFlow`` and I'm expecting it to behave like a proper ``Future[WebSocketUpgradeResponse]``

That would mean I shall be able to pass an ``onFailure`` partial function that would be invoked when TCP cannot connect ... **but that never happens instead!** Even though I attempt to connect to ``ws://127.0.0.1:9999/no/server/here``

I drilled down the source code and I suspect there could be something missing at 

https://github.com/akka/akka/blob/v2.4.8/akka-stream/src/main/scala/akka/stream/impl/io/TcpStages.scala#L208

Maybe the ``localAddressPromise`` should be broken on ``CommandFailed`` (in addition to failStage) as there won't be any local address to make the promise succeed because TCP cannot connect a server which does not exists.

```
msg match {
  case CommandFailed(cmd) => 
    val cause = new StreamTcpException(s"Tcp command [$cmd] failed")
    localAddressPromise.failure(cause)
    failStage(cause)
```

### PR DESCRIPTION
**Pull request by johanandren**
Fixes #21051

### CODE REVIEW

File: akka-stream-tests/src/test/scala/akka/stream/io/TcpSpec.scala [hunk 1]
```diff
@@ -76,6 +79,18 @@ class TcpSpec extends AkkaSpec("akka.stream.materializer.subscription-timeout.ti
...
+      val tcpWriteProbe = new TcpWriteProbe()
+      val future = Source.fromPublisher(tcpWriteProbe.publisherProbe)
+        .viaMat(Tcp().outgoingConnection(InetSocketAddress.createUnresolved("example.com", 666), connectTimeout = 1.second))(Keep.right)
+        .toMat(Sink.ignore)(Keep.left)
+        .run()
+
+      whenReady(future.failed) { ex =>
+        ex.getMessage should === ("Connection failed..")
+      }
+    }
```
Comment (johanandren) [inline akka-stream-tests/src/test/scala/akka/stream/io/TcpSpec.scala]: This already worked but couldn't find any pre-existing test coverage

File: akka-http-core/src/main/scala/akka/http/impl/engine/ws/WebSocketClientBlueprint.scala [hunk 2]
```diff
@@ -111,6 +111,11 @@ object WebSocketClientBlueprint {
           override def onPull(): Unit = pull(in)
 
           setHandlers(in, out, this)
+
+          override def onUpstreamFailure(ex: Throwable): Unit = {
+            result.tryFailure(new RuntimeException("Connection failed..", ex))
```
Comment (johanandren) [inline akka-http-core/src/main/scala/akka/http/impl/engine/ws/WebSocketClientBlueprint.scala]: After connection has succeeded the result is already completed, so this only happens on connection.

File: akka-http-core/src/test/scala/akka/http/impl/engine/ws/WebSocketIntegrationSpec.scala [hunk 3]
```diff
@@ -196,4 +199,19 @@ class WebSocketIntegrationSpec extends AkkaSpec("akka.stream.materializer.debug.
 
   }
 
+  "A websocket client" should {
+    "fail the materialized future if the request fails" in {
+      val flow = Http().webSocketClientFlow(
+        WebSocketRequest("ws://127.0.0.1:65535/no/server/here"),
+        settings = ClientConnectionSettings(system).withConnectingTimeout(250.millis))
+
+      val future = Source.maybe[Message].viaMat(flow)(Keep.right).toMat(Sink.ignore)(Keep.left).run()
```
Comment (ktoso) [inline akka-http-core/src/test/scala/akka/http/impl/engine/ws/WebSocketIntegrationSpec.scala]: `to(Sink.ignore).run()`

**Comment by ktoso** [comment]
LGTM

**Pull request merged**